### PR TITLE
tidy: check comment lengths in rust sources

### LINKED ISSUE #85726: tidy does not enforce comments only lines must be <= 100 characters long
`rustfmt` by default does not format comments so while it will reformat Rust code to be <= 100 characters per line, it does not do this for lines consisting only of a comment. As such `x.py tidy` does not report these lines and allows them to be merged into the repository without issue.

I tried enabling some `rustfmt` options to reformat comments but the results are suboptimal in many situations. For the time being, it would probably be best if `tidy` were taught to check for this issue itself.

See [this Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/tidy.20line.20lengths) for more details.

### PR DESCRIPTION
**Pull request by onur-ozkan**
Since rustfmt doesn't provide enough interface for detecting comment lengths and giving an error (also, we don't want to enable wrap_comments because autoformatting has false positives), I have enhanced tidy with small and simple additions to check comment lengths in the Rust sources.

Aiming to resolve #85726.

r? ghost (need to do some self-reviews as I blessed the whole source with rustfmt)

### CODE REVIEW

**Comment by onur-ozkan** [comment]
oh my lord, those labels...

File: src/tools/tidy/src/deps.rs [hunk 1]
```diff
@@ -54,10 +54,14 @@ pub(crate) const WORKSPACES: &[(&str, ExceptionList, Option<(&[&str], &[&str])>)
...
-    //("library/portable-simd", &[], None), // FIXME uncomment once rust-lang/portable-simd#363 has been synced back to the rust repo
-    //("library/stdarch", EXCEPTIONS_STDARCH, None), // FIXME uncomment once rust-lang/stdarch#1462 has been synced back to the rust repo
+    //("compiler/rustc_codegen_gcc", EXCEPTIONS_GCC, None), // FIXME uncomment once
+    //                                                         all deps are vendored
+    //("library/backtrace", &[], None), // FIXME uncomment once rust-lang/backtrace#562
+    //                                     has been synced back to the rust repo
+    //("library/portable-simd", &[], None), // FIXME uncomment once rust-lang/portable-simd#363
+    //                                         has been synced back to the rust repo
+    //("library/stdarch", EXCEPTIONS_STDARCH, None), // FIXME uncomment once rust-lang/stdarch#1462
+    //                                                  has been synced back to the rust repo
```
Comment (the8472) [inline src/tools/tidy/src/deps.rs:64]: So, I wanted to look at what logic tidy applies and I see this... that doesn't bode well for the rest of the PR

**Comment by the8472** [comment]
>  (also, we don't want to enable wrap_comments because autoformatting has false positives)

If we don't have good machine-applicable formatting then I'm opposed to this change because inserting text in a comment will trigger this and then require manually reflowing comment text.

File: src/tools/tidy/src/deps.rs (hunk 1 above)
Comment (onur-ozkan) [reply src/tools/tidy/src/deps.rs:64]: > So, I wanted to look at what logic tidy applies and I see this...

Tidy only checks, doesn't modify/bless.

That was done manually. How would you prefer to shorten that comment? There aren't many alternatives.

**Comment by fmease** [comment]
(Nice, triaging age-old <kbd>needs-triage-legacy</kbd> issues and adding the relevant labels seems to have a positive effect ❤️)

File: src/tools/tidy/src/deps.rs (hunk 1 above)
Comment (the8472) [reply src/tools/tidy/src/deps.rs:64]: I assume the line-length rule is supposed to improve readability. If it fails to achieve the goal then the block should be exempted from that.

**Comment by onur-ozkan** [comment]
> If we don't have good machine-applicable formatting 

I don't think we will ever have that.

> then I'm opposed to this change because inserting text in a comment will trigger this and then require manually reflowing comment text.

That's the main purpose. If you add comments (or modify existing ones) and when the length exceeds the maximum size, tidy will give errors on those lines to enforce you to fix it. I believe this should be manageable, as there shouldn't be many errors in each pull request.

**Comment by the8472** [comment]
> That's the main purpose. If you add comments (or modify existing ones) and when the length exceeds the maximum size, tidy will give errors on those lines to enforce you to fix it. 

Insert one word. Boom over the limit (which one might miss if one forgets to run tidy). PR fails. I need to go back and fix it and spill onto the next line. Which makes the next line-break awkward which means I have to reformat more lines....

I find this tiresome, unnecessary friction and of marginal benefit at best.

IMO linters should do one of these things:

* highlight real issues that require intervention
* automatically solve cosmetic issues
* get out of my way

Note that most comments already _aren't_ excessively long without the rule.

File: compiler/rustc_abi/src/lib.rs [hunk 2]
```diff
@@ -1152,7 +1152,6 @@ pub enum FieldsShape<FieldIdx: Idx> {
         /// it is easier to operate on the inverse mapping (that is,
         /// from memory order to source order), and that is usually
         /// named `inverse_memory_index`.
-        ///
```
Comment (Nadrieril) [inline compiler/rustc_abi/src/lib.rs:1155]: I think we should keep manually-inserted empty comment lines

File: compiler/rustc_pattern_analysis/src/usefulness.rs [hunk 3]
```diff
@@ -102,9 +103,9 @@
 //! of whether a constructor is matched by another one is answered by
 //! [`Constructor::is_covered_by`].
 //!
-//! Note 1: variable bindings (like the `x` in `Some(x)`) match anything, so we treat them as wildcards.
-//! Note 2: this only applies to matcheable values. For example a value of type `Rc<u64>` can't be
-//! deconstructed that way.
+//! Note 1: variable bindings (like the `x` in `Some(x)`) match anything, so we treat them as
+//! wildcards. Note 2: this only applies to matcheable values. For example a value of type `Rc<u64>`
```
Comment (Nadrieril) [inline compiler/rustc_pattern_analysis/src/usefulness.rs:107]: `Note 2` was intentionally on a new line

File: compiler/rustc_pattern_analysis/src/usefulness.rs [hunk 4]
```diff
@@ -20,7 +20,7 @@
 //! the following:
 //! - a set of values that match none of the patterns (if any),
 //! - for each subpattern (taking into account or-patterns), whether removing it would change
-//!     anything about how the match executes, i.e. whether it is useful/not redundant.
```
Comment (Nadrieril) [inline compiler/rustc_pattern_analysis/src/usefulness.rs:23]: That extra spacing was inserted on purpose for readability

File: compiler/rustc_abi/src/lib.rs (hunk 2 above)
Comment (onur-ozkan) [reply compiler/rustc_abi/src/lib.rs:1155]: Yeah that's why I made assigned ghost, I will revert changes like this.

**Comment by Nadrieril** [comment]
I'm ok with tidy checking comment lengths (I thought it already did in fact, so I've been reflowing comments every time), but I'm not a fan of fixing everything automatically like this because it's not reliable enough.

If I were you I would put all of these 700 files to a list of exceptions in tidy, and either ignore them forever (not that bad), or ask people to go through them piece-by-piece. Unless you do intend to sort through all that by hand in which case I bow to you

File: compiler/rustc_pattern_analysis/src/usefulness.rs (hunk 4 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/usefulness.rs:23]: I could accept the argument that this was not in fact more readable x)

File: compiler/rustc_codegen_cranelift/src/abi/mod.rs [hunk 5]
```diff
@@ -638,8 +639,8 @@ pub(crate) fn codegen_drop<'tcx>(
                 //
                 // WE CAN CONVERT THIS INTO THE ABOVE LOGIC BY DOING
                 //
-                // data = &(*args[0]).0    // gives a pointer to Data above (really the same pointer)
-                // vtable = (*args[0]).1   // loads the vtable out
+                // data = &(*args[0]).0    // gives a pointer to Data above (really the same
+                // pointer) vtable = (*args[0]).1   // loads the vtable out
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/src/abi/mod.rs:643]: This wrapping is wrong.

File: compiler/rustc_codegen_cranelift/src/base.rs [hunk 6]
```diff
@@ -502,7 +502,8 @@ fn codegen_stmt<'tcx>(
     fx.set_debug_loc(stmt.source_info);
 
     match &stmt.kind {
-        StatementKind::StorageLive(..) | StatementKind::StorageDead(..) => {} // Those are not very useful
+        StatementKind::StorageLive(..) | StatementKind::StorageDead(..) => {} /* Those are not */
+        // very useful
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/src/base.rs:506]: Putting the entire comment before the match arms looks much better.

**Comment by ChrisDenton** [comment]
This affects over 6k lines which maybe suggests to me that people are comfortable with at least slightly longer lines for comments than this PR allows. Would it be worth experimenting with increasing the limit for comments, at least to start with?

**Comment by Mark-Simulacrum** [comment]
I would agree with @bjorn3 here - people do a pretty good job of not adding very long comments as-is, I've not seen any significant PR discussion of "please reflow to fit X characters" (where a lint can help avoid review cycles), and it seems like comments are naturally a place where longer lengths may give more readability rather than less (e.g., diagrams, URLs, etc - I find limiting how these can be placed mostly unhelpful when I encounter it in other systems).

**Comment by onur-ozkan** [comment]
Closed the linked issue with https://github.com/rust-lang/rust/issues/85726#issuecomment-1913354938 and now closing this PR too.

**Pull request closed without merging**